Run quickcheck after counterexample was successful

evnu opened the pull request:
Currently, when a counterexample (#10) was found and it did not fail, no further tests are performed until the test is re-executed at a later point in time. This may come as a surprise to users, as they probably expect that the property is checked completely on each run. This changeset changes this behaviour by running quickcheck if the counterexample did not fail the test.

Note that this might not capture the intend of #10, so feel free to close this PR if the current behaviour is the intended behaviour.

alfert commented:
That's an interesting idea. The very basic idea of #10 is to focus on the failing property and expects that you work on that property only until it works (which is good, but you cannot guarantee that way of working, of course. At least, I do not follow it in all consequence). In `StreamData`, they use the ability of `ExUnit` to infer which tests must be rerun because some called modules are modified after the last test.  
Your idea tests all properties automatically, after the failing property succeeds again. This also ensures that all other modified implementation are checked. 

What do you think is a more usable approach and why? These features are all about usability and I am not sure what is best - as a default and as a configurable command line option.

evnu commented:
> What do you think is a more usable approach and why?

I know that I stumbled over forgetting to re-run the property after fixing a failing example. Example: I fixed the counter example and proceeded with running the complete test suite again, only to encounter another example failing with that property. At that point in time, I thought I had already fixed the property and forgot that it was only a single case which was fixed.

I do believe that both approaches have merit. Usually when fixing a property, I am focused on fixing only this one item and only run it. So the expectation that I am working only on that property usually applies.

I think I would be fine with a configuration option. Adding a command line option would work and allows adding a section to the README indicating how counterexamples work and how to change the behaviour. Additionally, when running a counterexample successfully to completion, a message could be written indicating that the property should be re-run to get a complete sample, or to use the command line flag in the future if that is what the user would prefer.

evnu commented:
@alfert Did you have time to consider this PR again? I again stumbled over this problem today.

alfert commented:
Let's go for it! But we need some documentation of this behaviour and the conflict must be resolved (besides a rebase, of course). Could you please take care of this? I will merge then.

evnu commented:
I will take care of it.

evnu commented:
I rebased, introduced a small helper function to refactor the identical calls of `PropCheck.quickcheck/2`, and added a paragraph to the property documentation in 21ecd52. Is the documentation enough?

alfert commented:
Looks good, thanks! I will publish a release today.
